src/matcher/mod.rs

```diff
@@ -0,0 +1,237 @@
+use std::{fmt, io, ops, u64};
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct Match {
+    start: usize,
+    end: usize,
+}
+
+impl Match {
+    #[inline]
+    pub fn new(start: usize, end: usize) -> Match {
+        assert!(start <= end);
+        Match { start, end }
+    }
+
+    #[inline]
+    pub fn zero(offset: usize) -> Match {
+        Match {
+            start: offset,
+            end: offset,
+        }
+    }
+
+    #[inline]
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    #[inline]
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    #[inline]
+    pub fn with_start(&self, start: usize) -> Match {
+        assert!(start <= self.end, "{} is not <= {}", start, self.end);
+        Match { start, ..*self }
+    }
+
+    #[inline]
+    pub fn with_end(&self, end: usize) -> Match {
+        assert!(self.start <= end, "{} is not <= {}", self.start, end);
+        Match { end, ..*self }
+    }
+
+    #[inline]
+    pub fn with_end_if_extends(&self, end: usize) -> Match {
```

Comment:
I added this since it was assuming somewhere that when it was "lumping" multiple matches together into a single range of lines the newer "lumpee"/match would end after the previous one(s) but that invariant wasn't actually true in tree-sitter world